packages/@lwc/engine-core/src/patches/detect-synthetic-cross-root-aria.ts

```diff
@@ -0,0 +1,183 @@
+    // shadow for a violation to be detected.
+    let vm: VM | undefined = getAssociatedVMIfPresent((source.getRootNode() as ShadowRoot).host);
+    if (isUndefined(vm)) {
+        vm = getAssociatedVMIfPresent((target.getRootNode() as ShadowRoot).host);
+    }
+    if (isUndefined(vm)) {
+        // vm should never be undefined here, but just to be safe, bail out and don't report
+        return;
+    }
+    report(ReportId.CrossRootAriaInSyntheticShadow, vm);
+    if (process.env.NODE_ENV !== 'production') {
+        logWarn(
+            `Element <${source.tagName.toLowerCase()}> uses attribute "${attrName}" to reference element ` +
+                `<${target.tagName.toLowerCase()}>, which is not in the same shadow root. This will break in native shadow DOM. ` +
+                `For details, see: https://lwc.dev/guide/accessibility#link-ids-and-aria-attributes-from-different-templates`,
+            vm
+        );
+    }
+}
+
+function parseIdRefAttributeValue(attrValue: any): string[] {
+    // split on whitespace and skip empty strings after splitting
+    return isString(attrValue) ? ArrayFilter.call(StringSplit.call(attrValue, /\s+/), Boolean) : [];
+}
+
+function detectSyntheticCrossRootAria(elm: Element, attrName: string, attrValue: any) {
+    const root = elm.getRootNode();
+    if (!isSyntheticShadowRootInstance(root)) {
+        return;
+    }
+
+    if (attrName === 'id') {
+        // elm is the target, find the source
+        if (!isString(attrValue) || attrValue.length === 0) {
+            // if our id is null or empty, nobody can reference us
+            return;
+        }
+        for (const idRefAttrName of ID_REFERENCING_ATTRIBUTES_SET) {
+            // Query all global elements with this attribute. The attribute selector syntax `~=` is for values
+            // that reference multiple IDs, separated by whitespace.
+            const query = `[${idRefAttrName}~="${CSS.escape(attrValue)}"]`;
+            const sourceElements = querySelectorAll.call(document, query);
+            for (let i = 0; i < sourceElements.length; i++) {
+                const sourceElement = sourceElements[i];
+                const sourceRoot = sourceElement.getRootNode();
+                if (sourceRoot !== root) {
+                    reportViolation(sourceElement, elm, idRefAttrName);
+                    break;
+                }
+            }
+        }
+    } else {
+        // elm is the source, find the target
+        const ids = parseIdRefAttributeValue(attrValue);
+        for (const id of ids) {
+            const target = getElementById.call(document, id);
+            if (!isNull(target)) {
+                const targetRoot = target.getRootNode();
+                if (targetRoot !== root) {
+                    // target element's shadow root is not the same as ours
+                    reportViolation(elm, target, attrName);
+                }
+            }
+        }
+    }
+}
```

Comment:
I was concerned this logic may be expensive, but I ran a quick test on CPM Record Home Inline Edit, and it only contributed 4ms to EPT even though there were hundreds of violations.

Still, we may want to use a gate in lwc-platform to turn the reporting off.